src/models/room_member.rs

```diff
@@ -0,0 +1,122 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::User;
+use crate::api::r0 as api;
+use crate::events::collections::all::{Event, RoomEvent, StateEvent};
+use crate::events::room::{
+    aliases::AliasesEvent,
+    canonical_alias::CanonicalAliasEvent,
+    member::{MemberEvent, MemberEventContent, MembershipChange, MembershipState},
+    name::NameEvent,
+    power_levels::PowerLevelsEvent,
+};
+use crate::events::EventResult;
+use crate::identifiers::{RoomAliasId, UserId};
+use crate::session::Session;
+
+use js_int::{Int, UInt};
+#[cfg(feature = "encryption")]
+use tokio::sync::Mutex;
+
+#[cfg(feature = "encryption")]
+use crate::crypto::{OlmMachine, OneTimeKeys};
+#[cfg(feature = "encryption")]
+use ruma_client_api::r0::keys::{upload_keys::Response as KeysUploadResponse, DeviceKeys};
+
+#[derive(Debug)]
+/// A Matrix room member.
+pub struct RoomMember {
+    /// The unique mxid of the user.
+    pub user_id: UserId,
+    /// The unique id of the room.
+    pub room_id: Option<String>,
+    /// If the member is typing.
+    pub typing: Option<bool>,
+    /// The user data for this room member.
+    pub user: User,
+    /// The users power level.
+    pub power_level: Option<Int>,
+    /// The normalized power level of this `RoomMember` (0-100).
+    pub power_level_norm: Option<Int>,
+    /// The `MembershipState` of this `RoomMember`.
+    pub membership: MembershipState,
+    /// The human readable name of this room member.
+    pub name: String,
+    /// The events that created the state of this room member.
+    pub events: Vec<Event>,
+}
+
+impl RoomMember {
+    pub fn new(event: &MemberEvent) -> Self {
+        let user = User::new(event);
+        Self {
+            room_id: event.room_id.as_ref().map(|id| id.to_string()),
+            user_id: event.sender.clone(),
+            typing: None,
+            user,
+            power_level: None,
+            power_level_norm: None,
+            membership: event.content.membership,
+            // TODO should this be `sender` ??
+            name: event.state_key.clone(),
+            events: vec![Event::RoomMember(event.clone())],
+        }
+    }
+
+    pub fn update_member(&mut self, event: &MemberEvent) -> bool {
+        use MembershipChange::*;
+
+        match event.membership_change() {
+            ProfileChanged => {
+                self.user.display_name = event.content.displayname.clone();
+                self.user.avatar_url = event.content.avatar_url.clone();
+                true
+            }
+            Banned | Kicked | KickedAndBanned | InvitationRejected | InvitationRevoked | Left
+            | Unbanned | Joined | Invited => {
+                self.membership = event.content.membership;
+                true
+            }
+            NotImplemented => false,
+            None => false,
+            // TODO should this be handled somehow ??
```

Comment:
I don't think we can do anything sensible here except ignore the event completely. Only a buggy or malicious server would send us such an event.